FINERACT-1747: add generic datatable query

ezolnbl-dpc opened the pull request:
## Description

see: https://issues.apache.org/jira/browse/FINERACT-1747


## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

hmg65 committed:
refactor: add lombok to emailcampaigns module

logoutdhaval committed:
FINERACT-1748: Loan prepayment issue after advance payment

Commit:
Chargeback Loan Transaction

renovate-bot committed:
chore(deps): update actions/stale digest to 99b6c70

dependabot[bot] committed:
chore(deps): bump actions/setup-node from 3.4.1 to 3.5.0

Bumps [actions/setup-node](https://github.com/actions/setup-node) from 3.4.1 to 3.5.0.
- [Release notes](https://github.com/actions/setup-node/releases)
- [Commits](https://github.com/actions/setup-node/compare/2fddd8803e2f5c9604345a0b591c3020ee971a93...969bd2663942d722d85b6a8626225850c2f7be4b)

---
updated-dependencies:
- dependency-name: actions/setup-node
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

ruchiD committed:
FINERACT-1694 Categorization for Events

Commit:
FINERACT-1747: add generic datatable query

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java:
```diff
@@ -183,6 +185,44 @@ public DatatableData retrieveDatatable(final String datatable) {
+    @Override
+    public List<JsonObject> queryDataTable(String datatable, String columnFilter, String valueFilter, String resultColumns) {
+        Arrays.asList(datatable, columnFilter, valueFilter, resultColumns).forEach(SQLInjectionValidator::validateDynamicQuery);
+        if(columnFilter == null || valueFilter == null) {
```
I think it would be better to have a validation error if the filter column or filter value is not provided.
The same validation should also apply for the resultColumns.

ezolnbl-dpc replied:
changing the code accordingly

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/utils/SQLInjectionValidator.java:
```diff
@@ -60,99 +60,72 @@ public static void validateSQLInput(final String sqlSearch) {
     }
 
-    public static void validateAdhocQuery(final String sqlSearch) {
+    public static void validateDynamicQuery(final String sqlSearch) {
```
It looks really similar like "validateSQLInput" method. Can we reuse that logic here?

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/utils/SQLInjectionValidator.java:
```diff
@@ -60,99 +60,72 @@ public static void validateSQLInput(final String sqlSearch) {
+        String lowerCaseSQL = sqlSearch.toLowerCase();
         for (String ddl : DDL_COMMANDS) {
-            if (lowerCaseSQL.startsWith(ddl)) {
+            if (ddl.equals(lowerCaseSQL)) {
```
I am not sure the "equals" are enough. I think the values cannot contains any of the restricted commands. 
e.g.: what if the "sqlSearch" value is ` 1; DROP TABLE m_loan; SELECT ` ?

adamsaghy commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java:
```diff
@@ -183,6 +185,44 @@ public DatatableData retrieveDatatable(final String datatable) {
+        }
+
+        List<String> resultColumnNames = resultColumns == null ? Collections.emptyList() : Stream.of(resultColumns.split(",")).toList();
+        final String sql = "select " + (resultColumns == null ? "" : resultColumns.trim()) + " from " + datatable + " where "
```
Please validate the provided filter column and resultColumns by the datatable metadata informations.
e.g.: this.genericDataService.fillResultsetColumnHeaders(dataTableName); -> it will give you the column headers metadata

ezolnbl-dpc replied:
changing the code accordingly

taskain7 committed:
[FINERACT-1678] business step configuration fix

logoutdhaval committed:
FINERACT-1755:withdrawal fee is not considered for interest calculation

renovate-bot committed:
chore(deps): update actions/setup-java digest to a18c333

daviestobialex committed:
Externalising OAuth Server URL 

Externalising the OAuth JWT URI for instances whereby the port and hostname will be different to what is currently configured, especially if you are using docker and you wish to use container names rather than IP or `localhost`

Commit:
Charge back Loan Transaction updates repayment schedule

ruchiD committed:
FINERACT-1694-Asynchronous-Message-OutChannel-Rename

adamsaghy commented on integration-tests/src/test/java/org/apache/fineract/integrationtests/DatatableIntegrationTest.java:
```diff
@@ -247,6 +252,15 @@ public void validateCreateReadDeleteDatatable() {
         assertEquals(datatableEntryMap.get("itsAString"), ((Map) updatedDatatableEntryResponse.get("changes")).get("itsAString"));
         assertEquals(datatableEntryMap.get("itsAText"), ((Map) updatedDatatableEntryResponse.get("changes")).get("itsAText"));
 
+        FineractClient fineract = FineractClient.builder().insecure(true).baseURL("https://localhost:8443/fineract-provider/api/v1/")
```
This can be replaced with the "IntegrationTest.fineract()".

ezolnbl-dpc replied:
replaced

adamsaghy reviewed (approved):
LGTM

adamsaghy commented on integration-tests/src/test/java/org/apache/fineract/integrationtests/DatatableIntegrationTest.java:
```diff
@@ -247,6 +252,15 @@ public void validateCreateReadDeleteDatatable() {
 
+        FineractClient fineract = FineractClient.builder().insecure(true).baseURL("https://localhost:8443/fineract-provider/api/v1/")
+                .tenant("default").basicAuth("mifos", "password").build();
+        String rows = Calls.ok(fineract.dataTables.queryValues(datatableName, "itsABoolean",
```
Please call a ".toLowerCase" on the "datatableName" variable. In postgres unfortunately the lookup not do it automatically.

ezolnbl-dpc replied:
fixed

adamsaghy reviewed (changes requested):
See my comments

taskain7 committed:
[FINERACT-1678] Loan accounts stayed locked business event

renovate-bot committed:
chore(deps): update all non-major dependencies

dependabot[bot] committed:
chore(deps): bump actions/checkout from 3.0.2 to 3.1.0

Bumps [actions/checkout](https://github.com/actions/checkout) from 3.0.2 to 3.1.0.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/2541b1294d2704b0964813337f33b291d3f8596b...93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

ruchiD committed:
FINERACT-1694-TenantIdentifier-ExternalEvents

adamsaghy commented on integration-tests/src/test/java/org/apache/fineract/integrationtests/DatatableIntegrationTest.java:
```diff
@@ -240,24 +248,28 @@ public void validateCreateReadDeleteDatatable() {
+        assertEquals(datatableEntryMap.get(itsAString), ((Map) updatedDatatableEntryResponse.get("changes")).get(itsAString));
+        assertEquals(datatableEntryMap.get(itsAText), ((Map) updatedDatatableEntryResponse.get("changes")).get(itsAText));
+
+        Calls.ok(fineract().dataTables.queryValues(datatableName, itsANumber, datatableEntryMap.get(itsANumber).toString(), itsANumber));
```
Please test the returning values whether they are correct

renovate-bot committed:
chore(deps): update all non-major dependencies

renovate-bot committed:
fix(deps): update non-major

dependabot[bot] committed:
chore(deps): bump actions/stale from 5.2.0 to 6.0.1

Bumps [actions/stale](https://github.com/actions/stale) from 5.2.0 to 6.0.1.
- [Release notes](https://github.com/actions/stale/releases)
- [Changelog](https://github.com/actions/stale/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/stale/compare/99b6c709598e2b0d0841cd037aaf1ba07a4410bd...5ebf00ea0e4c1561e9b43a292ed34424fb1d4578)

---
updated-dependencies:
- dependency-name: actions/stale
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

ruchiD committed:
FINERACT-1694-Regular-Event-Purging

galovics committed:
FINERACT-1724: Enhancements on batch API support for adjusting a loan transaction

tapaswenipathak committed:
Refine documentation for migration of database from MySQL to PostgreSQL including the migration from Flyway to Liquibase for database migration scripts

wkigenyi committed:
FINERACT-1724: Correct descriptions and summary on DELETE and PUT (#2458)

FINERACT-1724: Correct descriptions and summary on DELETE and PUT

vidakovic committed:
PS-15: Synchronize the openmf develop branch with upstream

vidakovic committed:
PS-15: Revert

galovics committed:
FINERACT-1724: Swagger fixes for PostLoansLoanIdTransactionsTransactionIdRequest

hmg65 committed:
refactor: add lombok to entityaccess module

hmg65 committed:
refactor: add lombok to jobs module

hmg65 committed:
refactor: add lombok to reportmailingjob module
